src/transaction/log_replication_atomic.cpp

```diff
@@ -367,4 +368,49 @@ namespace cublog
 	  }
       }
   }
+
+  void
+  atomic_replicator::release_all_locks_for_ddl (cubthread::entry &thread_entry, const TRANID trid)
+  {
+    assert (m_locked_classes.count (trid) > 0);
+
+    auto [begin, end] = m_locked_classes.equal_range (trid);
+    for (auto it = begin; it != end; it++)
+      {
+	lock_unlock_object_and_cleanup (&thread_entry, & (it->second), oid_Root_class_oid, SCH_M_LOCK);
+      }
+
+    m_locked_classes.erase (trid);
+  }
+
+  void
+  atomic_replicator::acquire_lock_for_ddl (cubthread::entry &thread_entry, const TRANID trid, const OID *classoid)
+  {
+    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
+
+    /* TODO:
+     * If a PTS read transaction holds a lock for an extended period without releasing it for the same class,
+     * the replicator could wait too long to acquire the lock.
+     * In such cases, there might be an introduction of a mechanism to abort read transaction that holds lock.
+     */
+
+    if (lock_object (&thread_entry, classoid, oid_Root_class_oid, SCH_M_LOCK, LK_UNCOND_LOCK) != LK_GRANTED)
+      {
+	assert_release (false);
+      }
+
+    m_locked_classes.emplace (trid, *classoid);
+
+#if !defined (NDEBUG)
+    /* there will be no duplicated object in the map */
+    auto [begin, end] = m_locked_classes.equal_range (trid);
+    for (auto it = begin; it != end; it++)
+      {
+	if (std::count (begin, end, *it) > 1)
+	  {
+	    assert (false);
+	  }
+      }
+#endif
```

Comment:
I'd check the duplication only for <trid, *classoid>. It seems to check all classoid for a trid.
```suggestion
#if !defined (NDEBUG)
    /* there will be no duplicated object in the map */
    auto [begin, end] = m_locked_classes.equal_range (trid);
    for (auto it = begin; it != end; it++) // or count_if()?
      {
          assert (it.second != *classoid);
      }
#endif
    m_locked_classes.emplace (trid, *classoid);
```

Reply:
Oh..?! this is much simpler, thanks